src/js/stores/search.js

```diff
@@ -0,0 +1,212 @@
+         */
+        getHeaders: function (id) {
+            return this._registeredSearches[id].searchHeaders;
+        },
+
+        /**
+         * Perform action for item
+         *
+         * @param {string} id Search module to use
+         * @param {string} itemID Search item to handle
+         */
+        handleExecute: function (id, itemID) {
+            return this._defaultHandleExecute(id, itemID);
+        },
+
+        /**
+         * Add search module and find initial items, headers, filters
+         *
+         * @param {string} id The search module ID
+         * @param {Array.<string>} types What to search for
+         */
+        registerSearch: function (id, types) {
+            this._registeredSearches[id] = { "searchHeaders": types };
+            this._updateSearchItems(id);
+        },
+
+        /**
+         * Add search type
+         *
+         * @param {object} payload
+         */
+        _registerSearchType: function (payload) {
+            this._registeredSearchTypes[payload.type] = {
+                "getOptions": payload.getOptions,
+                "filters": payload.filters,
+                "handleExecute": payload.handleExecute
+            };
+        },
+
+        /**
+         * Add all possible options for each of the search headers for the module
+         *
+         * @param {string} id The search module ID
+         */
+        _updateSearchItems: function (id) {
+            var search = this._registeredSearches[id],
+                types = search.searchHeaders;
+
+            if (types) {
+                search.searchItems = _.reduce(types, function (items, type) {
+                    return items.concat(this._getOptions(type));
+                }.bind(this), Immutable.List());
+
+                search.searchFilters = this._getFiltersByHeaders(types);
+            }
+        },
+
+        /**
+         * Call correct get options function
+         *
+         * @param {string} type
+         * @return {Immutable.List}
+         */
+        _getOptions: function (type) {
+            if (this._registeredSearchTypes[type]) {
+                return this._registeredSearchTypes[type].getOptions();
+            }
+            return Immutable.List();
+        },
+
+        /**
+         * Get subcategories for all the headers
+         *
+         * @param {Array.<string>} headers
+         * @return {Array.<Immutable.List>}
+         */
+        _getFiltersByHeaders: function (headers) {
+            return _.map(headers, function (header) {
+                if (this._registeredSearchTypes[header]) {
+                    return this._registeredSearchTypes[header].filters;
+                }
+            }.bind(this), []);
+        },
+
+        /**
+         * Perform default action for item
+         *
+         * @param {string} id Search module ID
+         * @param {string} itemID Selected search item ID
+         */
+        _defaultHandleExecute: function (id, itemID) {
```

Comment:
Per my comments above, I think you should be able to skip this forEach search and immediately dive into the action registered as a "handler" for a menu item, given you provide it on registration.